### PR TITLE
Merge 12-stable into master

### PR DESCRIPTION
**Pull request by jdmundrawala**
Things in 12-stable not in master (that merged cleanly)

cc @lamont-granquist @btm @chef/client-engineers

### CODE REVIEW

File: spec/support/shared/unit/execute_resource.rb [hunk 1]
```diff
@@ -86,6 +86,12 @@
     expect(@resource.path [ "woot" ]).to raise_error
   end
 
+  it "should accept an array for the execution path in Chef-12 and log deprecation message" do
+    expect(Chef::Log).to receive(:warn).at_least(:once)
+    @resource.path ["woot"]
+    expect(@resource.path).to eql(["woot"])
+  end
+
```
Comment (btm) [inline spec/support/shared/unit/execute_resource.rb]: 7f760f67 added this already, it's on line 79 above.

File: spec/unit/api_client/registration_spec.rb [hunk 2]
```diff
@@ -231,6 +231,28 @@
...
+      context 'with follow_client_key_symlink set to true' do
+        before do
+          Chef::Config[:follow_client_key_symlink] = true
+        end
+
+        it 'follows the symlink', :unix_only do
+          expect(registration.file_flags).to eq(File::CREAT|File::TRUNC|File::RDWR)
+        end
+      end
+    end
```
Comment (btm) [inline spec/unit/api_client/registration_spec.rb]: This is already in master as 9d072529e + 3630b9bae. You can see these above too.

File: spec/unit/application/client_spec.rb [hunk 3]
```diff
@@ -86,6 +86,18 @@
...
+        Chef::Config[:interval] = 600
+        allow(Chef::Platform).to receive(:windows?).and_return(true)
+      end
+
+      it "should not terminate" do
+        expect(Chef::Application).not_to receive(:fatal!)
+        @app.reconfigure
+      end
+    end
+
```
Comment (btm) [inline spec/unit/application/client_spec.rb]: This also already exists on master, just above, from 6e5cb19c.

File: spec/unit/knife/bootstrap_spec.rb [hunk 4]
```diff
@@ -79,6 +79,30 @@
...
+  end
+
+  context "with :bootstrap_template and :template_file cli options" do
+    let(:bootstrap_cli_options) { [ "--bootstrap-template", "my-template", "--template-file", "other-template" ] }
+
+    it "should select bootstrap template" do
+      expect(File.basename(knife.bootstrap_template)).to eq("my-template")
+    end
+  end
+
```
Comment (btm) [inline spec/unit/knife/bootstrap_spec.rb]: duplicate tests

File: spec/unit/provider/file/content_spec.rb [hunk 5]
```diff
@@ -98,6 +98,28 @@
...
+        Chef::Config[:file_staging_uses_destdir] = true
+        expect{content.tempfile}.to raise_error
+      end
+
+      it "returns a tempfile in the tempdir when :file_desployment_uses_destdir is not set" do
+        expect(content.tempfile.path.start_with?(Dir::tmpdir)).to be_true
+        expect(canonicalize_path(content.tempfile.path).start_with?(enclosing_directory)).to be_false
+      end
+    end
+
```
Comment (btm) [inline spec/unit/provider/file/content_spec.rb]: duplicate tests f5ec7e9c

**Comment by jdmundrawala** [comment]
this is a bad idea

**Pull request closed without merging**